gnmi/README.md

```diff
@@ -1,22 +1,59 @@
-is registered.
+Reconcilers are created by implementing the [Reconciler interface](https://pkg.go.dev/github.com/openconfig/lemming/gnmi/reconciler#Reconciler), and optionally using the a reconciler Builder to simplify the creation.
 
-## 2 Subscribing using ygnmi
+See the [fakedevice package](fakedevice/fakedevice.go) for examples of reconcilers.
 
-This method uses [`ygnmi`](https://github.com/openconfig/ygnmi) to receive
-updates from the central database.
+### Reconciliation
 
-See [testagent](testagent/testagent.go) for an example.
+Reconcilers must implement the the reconciler.Reconciler interface. The reconciler interface is intentionally minimum to minimize the burden of creating reconcilers and allow maximum flexibility when creating a reconciler.
+
+#### Reacting to Values
+
+Reconcilers can react to values by either calling a raw gNMI operation using the provided GNMIClient argument or by starting a Watch using ygnmi (optionally using WithWatch). ygnmi.Watch can be used monitoring the current state of a set of OpenConfig subtrees/leaves using gNMI.Subscribe.
+
+#### Updating Values
+
+Reconcilers can update values in the cache by sending raw gNMI set requests using the client or using ygnmi. The gnmiclient package contains special helpers that allow using ygnmi.Update/Replace/Delete on state paths. Generally, internal reconcilers should not be updating config paths as those are set by external users.
+
+NOTE: Currently, Set requests are required to be schema-compliant.
+NOTE: Replace/Update on non-leaf state values is not yet. See [issue](https://github.com/openconfig/lemming/issues/67).
+
+### Validation
+
+Reconcilers can validate intended config before it is applied to the cache. To do so, Reconcilers must return a non-empty set of path prefixes they are interested in validating through `Reconciler.ValidationPaths()`.
+For any SetRequest that modifies those paths, the Validate function is called on the global intended config.
+
+Note: this currently implemented as any reconciler that returns > 0 paths is called on every SetRequest.
+
+### Reconciler Builders
+
+The reconciler package provides a builder API for creating reconcilers with common use cases:
+
+```go
+rec := reconciler.NewTypedBuilder[string]("hostname").
+    WithWatch(hostnamePath.Config(), func(ctx context.Context, c *ygnmi.Client, v *ygnmi.Value[string]) error { // Watch hostname config: set hostname, and update state.
+        hostname, ok := v.Val()
+        if !ok {
+            resetHostName()
+        }
+        setHostName(hostName)
+        if _, err := gnmiclient.Update(ctx, c, hostnamePath.State(),hostname); err != nil {
+            return log.Warningf("Failed to set hostname: %v",err)
+        }
+        return ygnmi.Continue
+    }).Build()
+```
+
+Alternatively, a custom reconciler implementation matching the Reconciler interface can be created and registered on the gnmi server.
```

Comment:
Perhaps add a line afterwards on how these reconcilers can be registered onto the device.

Reply:
added earlier in doc